Add win32 build and replace stored external libs with builds from sources on all platforms

vitaly-krugl opened the pull request:
* The large number of "Files changed" is due to a very large number of obsoleted external lib files being deleted. The new build uses tar.gz and zip archives to build from sources.
* Visual Studio build doesn't work. It was never supported by nupic.core's CI. I tried adding it to appveyor.yml's matrix, but it fails, and will need TLC from someone. I commented out its variant in appveyor.yml's matrix so that it doesn't consume resources unnecessarily.
* Win32 build of nupic.core libraries completes and almost all tests pass, except a test in unit_tests.exe, which is most likely due to 32/64-bit portability issues. @scottpurdy will be investigating this test failure.
* Win64 builds and all tests pass
* Travis Linux gcc/clang and OS X clang: I added the balance of the unit nupic.core tests to .travis.yml that were originally excluded for some reason. All are building and passing now.
* Also fixes #887

vitaly-krugl commented:
Hmm, post-squashing, the linux GCC build died of segmentation fault:

```
Linking CXX executable capnp
cd /home/travis/build/numenta/nupic.core/build/scripts/ThirdParty/Build/CapnProto/src/capnp && /usr/bin/cmake -E cmake_link_script CMakeFiles/capnp_tool.dir/link.txt --verbose=1
/usr/bin/g++-4.8   -std=gnu++0x -Wall -Wextra -Wno-strict-aliasing -Wno-sign-compare -Wno-unused-parameter -pthread   -std=c++11 -Wall -fPIC -m64  -static-libgcc -Wextra -Wreturn-type -Wunused -Wno-unused-variable -Wno-unused-parameter -Wno-missing-field-initializers -Wno-deprecated-register -pipe  -mtune=generic -O2 -fuse-linker-plugin -flto-report -flto -fuse-ld=gold  -static-libstdc++ -DHAVE_UNISTD_H    CMakeFiles/capnp_tool.dir/compiler/module-loader.c++.o CMakeFiles/capnp_tool.dir/compiler/capnp.c++.o  -o capnp -rdynamic libcapnpc.a libcapnp.a ../kj/libkj.a 
WPA statistics
[WPA] GIMPLE type table is empty
[WPA] GIMPLE type hash table is empty
[WPA] GIMPLE canonical type table is empty
[WPA] GIMPLE canonical type hash table is empty
[WPA] # of input files: 33
[WPA] # of input cgraph nodes: 4969
[WPA] # of function bodies: 0
[WPA] # of output files: 31
[WPA] # of output symtab nodes: 11399
[WPA] # callgraph partitions: 31
[WPA] Compression: 7097054 input bytes, 24528558 uncompressed bytes (ratio: 3.456161)
[WPA] Size of mmap'd section decls: 2342100 bytes
[WPA] Size of mmap'd section function_body: 4440327 bytes
[WPA] Size of mmap'd section statics: 0 bytes
[WPA] Size of mmap'd section symtab: 0 bytes
[WPA] Size of mmap'd section refs: 15616 bytes
[WPA] Size of mmap'd section asm: 0 bytes
[WPA] Size of mmap'd section jmpfuncs: 75482 bytes
[WPA] Size of mmap'd section pureconst: 12251 bytes
[WPA] Size of mmap'd section reference: 0 bytes
[WPA] Size of mmap'd section symbol_nodes: 130520 bytes
[WPA] Size of mmap'd section opts: 0 bytes
[WPA] Size of mmap'd section cgraphopt: 0 bytes
[WPA] Size of mmap'd section inline: 80758 bytes
[WPA] Size of mmap'd section ipcp_trans: 0 bytes
/home/travis/build/numenta/nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/parse/common.h: In member function ‘operator()’:
/home/travis/build/numenta/nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/parse/common.h:639:7: internal compiler error: Segmentation fault
       operator()(Input& input) const {
       ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-4.8/README.Bugs> for instructions.
lto-wrapper: /usr/bin/g++-4.8 returned 1 exit status
/usr/bin/ld.gold: fatal error: lto-wrapper failed
collect2: error: ld returned 1 exit status
make[5]: *** [src/capnp/capnp] Error 1
make[5]: Leaving directory `/home/travis/build/numenta/nupic.core/build/scripts/ThirdParty/Build/CapnProto'
make[4]: *** [src/capnp/CMakeFiles/capnp_tool.dir/all] Error 2
make[4]: Leaving directory `/home/travis/build/numenta/nupic.core/build/scripts/ThirdParty/Build/CapnProto'
make[3]: *** [all] Error 2
make[3]: Leaving directory `/home/travis/build/numenta/nupic.core/build/scripts/ThirdParty/Build/CapnProto'
make[2]: *** [ThirdParty/Stamp/CapnProto/CapnProto-build] Error 2
make[2]: Leaving directory `/home/travis/build/numenta/nupic.core/build/scripts'
make[1]: *** [external/CMakeFiles/CapnProto.dir/all] Error 2
make[1]: Leaving directory `/home/travis/build/numenta/nupic.core/build/scripts'
make: *** [all] Error 2
```

vitaly-krugl commented:
Okay, I will need to debug that recurring compiler segmentation fault during capnproto build. A last-minute change that I made to the capnproto build is most likely to blame.

scottpurdy commented on CommonCompilerConfig.cmake:
```diff
@@ -0,0 +1,184 @@
+      -D_VARIADIC_MAX=10
+      -DNOMINMAX
+      -DCAPNP_LITE=1)
+  if(MSYS OR MINGW)
```
Rather than nesting this, I'd move it after the outer if statement.

scottpurdy commented on CommonCompilerConfig.cmake:
```diff
@@ -0,0 +1,184 @@
+
+
+#
+# Set linker (ld)
```
Did we do something like this before? Why do we need to specify a linker?

vitaly-krugl replied:
@scottpurdy: The `ld.gold` logic came directly from the old src/CMakeFiles.txt. Unfortunately, the comment didn't explain why ld.gold is so desirable that the code is going to the trouble for it. I'd be happy to get rid of it, if you think we should.

breznak replied:
I think the linker is needed for the LTO optimizations, which are/can be used under source builds. ld.gold is better/only worked at the time.

scottpurdy commented:
In general, I am a little wary of the common flags that are applied to externals as well. Compiler settings sometimes vary by project so we may need to use a different set for some of the externals (I think).

scottpurdy commented on external/Apr1Lib.cmake:
```diff
@@ -0,0 +1,160 @@
+    #LOG 1
+
+    COMMAND
+        ${CMAKE_COMMAND} -E echo "Patching <SOURCE_DIR> via ${aprlib_patch_file}"
```
Is this necessary? It seems that this duplicates the `COMMENT`, which the documentation says is printed when the step executes.

scottpurdy commented on external/Apr1Lib.cmake:
```diff
@@ -0,0 +1,160 @@
+        DEPENDEES install
+        ALWAYS 1
+
+        # TODO ZZZ remove these
```
remove these...

scottpurdy commented on external/Apr1Lib.cmake:
```diff
@@ -0,0 +1,160 @@
+set(aprlib_install_prefix "${EP_BASE}/Install/Apr1StaticLib")
+set(aprlib_install_lib_dir "${aprlib_install_prefix}/lib")
+
+# Export directory of installed apr-1 lib headers to parent
```
These comments saying they are exported to the parent are confusing. There is a `PARENT_SCOPE` option for the `set` command that isn't used here (and isn't necessary depending on how CMakeLists.txt includes this). So perhaps reword / drop the parent comment?

scottpurdy commented on external/AprUtil1Lib.cmake:
```diff
@@ -0,0 +1,151 @@
+            make -f Makefile all
+
+        INSTALL_COMMAND
+            make -f Makefile install
```
How does this differ from the default? Isn't `-f Makefile` the default behavior?

vitaly-krugl replied:
Explicit better than implicit?

scottpurdy commented:
It looks like you are using a convention for cmake variables where you use lower_case for local variables and UPPER_CASE for "exported" variables. Is that right? Is that a convention you saw used elsewhere? I like it but haven't noticed it before.

vitaly-krugl commented:
@scottpurdy: regarding lowercase var convention in cmake files: yes, during this exercise, I had an opportunity to scan through a number of 3rd party cmake files, and this particularly appealed to me as a way to distinguish "local" from "global" variables in cmake. Of course, since many of the modules are imported via `include`, all of their variables still bleed into the parent, including the "internal" ones. It's unfortunate that there isn's some way to prevent this, as I am sure it causes some difficult-to-debug problems on occasion.

vitaly-krugl commented:
@scottpurdy, regarding
> In general, I am a little wary of the common flags that are applied to externals as well. Compiler settings sometimes vary by project so we may need to use a different set for some of the externals (I think).

This was a solution to a real problem. There are a number of settings that we need to have in common, and as I was introducing new external lib builds into nupic.core build, I was getting linker errors. For example, `-fPIC` is needed for producing position-independent code for DLLs, and all objects that will be linked into a DLL need to be compiled with that flag. Then, there are BITNESS flags, optimization flags, debugging-symbol generation flags, etc.  It was impossible to maintain consistency between modules without a way to share settings across modules, and do this correctly everywhere across compiler toolchains and OS platforms.

Also, from `man g++` on Ubuntu in the `-flto` section (but it has a more general connotation):

>
![screen shot 2016-03-29 at 6 27 35 pm](https://cloud.githubusercontent.com/assets/1441066/14129118/f8cfbe42-f5db-11e5-862f-4b4ba21af0db.png)

vitaly-krugl commented on external/CapnProto.cmake:
```diff
@@ -114,6 +122,7 @@ foreach (INCLUDE_DIR ${CAPNP_INCLUDE_DIRS})
   install(DIRECTORY ${INCLUDE_DIR}/capnp
```
@scottpurdy It's confusing that these same INC and LIB installation steps are duplicated in src/CMakeLists.txt. Are they in fact accomplishing anything different and useful?

scottpurdy replied:
Yes they are different. The externals are configured to install into <build_dir>/ThirdParty/Install (or something like that, it's an ExternalProject convention). This is great for most cases where the installed artifacts are only needed for the build of nupic.core. But for capnp we want to also include the capnp install artifacts in the installation of nupic.core and that is what is specified in src/CMakeLists.txt (which is responsible for what gets installed to the user-specified install location).

vitaly-krugl replied:
@scottpurdy: I looked at this once more and it was definitely worth a second look. When I run `make install` on my MacBookPro, there are two attempts to install the capnp files under the following headings:

1. `Installing: /Users/vkruglikov/nta/nupic.core/build/release/include//capnp`
2. `Installing: /Users/vkruglikov/nta/nupic.core/build/release/include/capnp`

Note that the headings only differ in two slashes versus a single slash before the last component. This corresponds precisely to the two duplicated sets of install commands that only vary by the presence/absence of an extra slash:

From CapnProto.cmake:
```
foreach (INCLUDE_DIR ${CAPNP_INCLUDE_DIRS})
  install(DIRECTORY ${INCLUDE_DIR}/kj
          DESTINATION include/)
  install(DIRECTORY ${INCLUDE_DIR}/capnp
          DESTINATION include/)
endforeach ()

install(FILES ${CAPNP_LINK_LIBRARIES}
        DESTINATION lib/)
```

From src/CMakeFiles.txt:

```
install(FILES ${CAPNP_LINK_LIBRARIES}
        DESTINATION lib)

foreach(directory ${CAPNP_INCLUDE_DIRS})
  install(DIRECTORY "${directory}/capnp"
          DESTINATION include)
  install(DIRECTORY "${directory}/kj"
          DESTINATION include)
endforeach()

```

vitaly-krugl commented:
Hi @scottpurdy, I added a few more commits following a98d373fd4b563af99ce15f7ee16252158469079. The first one is a bit more involved, and the rest are tiny. Please take a look. I also addressed your earlier feedback starting at a98d373fd4b563af99ce15f7ee16252158469079.

breznak commented on CommonCompilerConfig.cmake:
```diff
@@ -0,0 +1,239 @@
+  # LLVM Clang / Gnu GCC
+  set(COMMON_CXX_FLAGS_UNOPTIMIZED "${COMMON_CXX_FLAGS_UNOPTIMIZED} ${stdlib_cxx} -std=c++11")
+
+  set(shared_compile_flags "-m${BITNESS} ${stdlib_common} -Wextra -Wreturn-type -Wunused -Wno-unused-variable -Wno-unused-parameter -Wno-missing-field-initializers")
```
if `shared_compile_flags` are the flags for building libraries, we probably could drop all the `W*` flags.

scottpurdy replied:
Yes, we can't enforce arbitrary warnings on externals. I'm also not sure that we want to build all externals as C++11. The only things we should have in the common set are those that really need to be the same for our code and all externals, which should be very few.

vitaly-krugl replied:
I am open to explicit proposals that would improve things. On the other hand, the current common settings produce a functional build already ...

breznak commented on ci/travis/before_install-linux.sh:
```diff
@@ -41,8 +41,16 @@ export PATH=$HOME/.local/bin:$PATH
+echo "WHEREIS PIP: $(whereis pip)"
+/usr/local/bin/pip --version
+/usr/local/bin/pip2.7 --version
+echo "PATH=$PATH"
```
still a problem, or forgotten debug?

vitaly-krugl replied:
@breznak, It's still a problem that needs to be solved. pip complains that the current version 6.x and the latest is 8.x. The `--upgrade` downloads the 8.x wheel and "installs", then prints "successfully installed 6.x" (forgot the exact x). Any idea how to really fix this?

breznak commented on external/CapnProto.cmake:
```diff
@@ -19,71 +19,87 @@
+
+
+set(capnproto_lib_url "${REPOSITORY_DIR}/external/common/share/capnproto/capnproto-c++-0.5.3.tar.gz")
+set(capnproto_win32_tools_url "${REPOSITORY_DIR}/external/common/share/capnproto/capnproto-c++-win32-0.5.3.zip")
```
use a $version in these?

vitaly-krugl replied:
@breznak, are you recommending defining a variable `set(version "0.5.3") and then using it like this:

```
set(capnproto_lib_url "${REPOSITORY_DIR}/external/common/share/capnproto/capnproto-c++-${version}.tar.gz")
```

If so, it's not worth it, since the rest of the naming scheme is not consistent in the capnp's other tagged releases in their github.

breznak commented:
Nice to see so many `external/` headers removed :+1: 
- [ ] a problem I see is that it seems this PR removes the functionality of `Werror` which is an (easy) step backwards. 
Btw, could you tell me which flags are used for "common lib" compilation and which for "common our src" (as we discussed in the issue)?

scottpurdy commented:
@vitaly-krugl there are a few comments waiting on you

vitaly-krugl commented:
@scottpurdy, I was at Strata all day today disseminating our info at the booth, so just getting back to coding.

vitaly-krugl commented:
@breznak, regarding

> a problem I see is that it seems this PR removes the functionality of Werror which is an (easy) step backwards.

`-Werror` is the kind of setting that woks for our own sources, but doesn't work with some of the externals, so it's combined with the common cxx optimized settings in src/CMakeLists.txt and applied only to our own sources.

vitaly-krugl commented:
@breznak, regarding

> Btw, could you tell me which flags are used for "common lib" compilation and which for "common our src" (as we discussed in the issue)?

The build is functional with the following settings on all supported CI platforms in both build and debug build types:

From memory (may double-check in https://github.com/vitaly-krugl/nupic.core/tree/add-win32-build/external)
* src: common cxx optimized + `-include cmath` + `-Werror`
* swig-generated sources: common cxx optimized + `-include cmath`
* apr and apr-util: common c optimized
* yamlcpp: common cxx optimized
* yaml: common c optimized
* zlib: common c optimized
* capnproto: common cxx unoptimized (conflicted with the optimized settings, so downgraded to unoptimized)

vitaly-krugl commented:
@scottpurdy, I responded to all outstanding comments. Please advise on further action, if any. I would like to turn attention to the nupic side of things ASAP to validate what we have done here end-to-end.

breznak commented:
Thanks for explanations @vitaly-krugl ! :+1: 

> `-Werror` ... so it's combined with the common cxx optimized settings in src/CMakeLists.txt and applied only to our own sources.

Great then! I got confused by the resemblance of the new `CommonCompilerConfig.cmake` to CMakeLists and thought Werror removed.

scottpurdy commented:
👍 (still work to do but better to do in follow up PRs)
